internal: migrate to vscode.FileSystem API

wxb1ank opened the pull request:
I encountered an error where `bootstrap()` attempts to create a directory with the path `C:\C:\...`. I couldn't find this reported anywhere else. Using the `vscode.FileSystem` API instead of the `fs` one works here. I assume the latter automatically prepends `C:\` to paths whereas the former does not. I don't know if this suggests `vscode.FileSystem` should be used in more places for consistency.

matklad commented:
yeah, ideally we should never use bare fs are there other places we might want to adjust this?

wxb1ank commented:
Well, it looks like `fs` is used in `main.ts`, `net.ts`, and `toolchain.ts`, at least. I don't immediately see any situations where `vscode.FileSystem` would be insufficient. I'm AFK at the moment but can look into replacing those calls later.

I guess the question at this point is which API is more stable and if the VS Code one has drawbacks I'm not aware of. I admit I'm not very familiar with Node/TypeScript, so please correct me if I'm mistaken on anything :)

wxb1ank commented:
I replaced some cases of low-hanging fruit using `fs` with `vscode.FileSystem` equivalents. As the VS Code API uses `vscode.Uri` for file paths (and not plain old `string`s), I also introduced some `Uri.file()` and `Uri.joinPath()` code to avoid Node's `path` module. I would have liked to completely eliminate the need for `path` and `fs`, but unfortunately:

- `vscode.Uri` is kind of limited, and `vscode.FileSystem` lacks some functionality (e.g., `createWriteStream()`) of `fs`
- `vscode.Uri.file()` prepends `/` to the given path, which is undesirable when referencing executables in the PATH

yume-chan commented:
The root cause is that in 95c51d8f1df1efdb2e98b7717544a6db8cad14e7

```diff
- this.globalStoragePath = ctx.globalStoragePath;
+ this.globalStoragePath = ctx.globalStorageUri.path;
```

`Uri#path` returns `/C:/Users/XXX` on Windows. The correct property should be `Url#fsPath` (https://code.visualstudio.com/api/references/vscode-api#details-418), it returns `c:\Users\XXX`

I'm too familiar with new APIs like `vscode.FileSystem` thus don't know which one is better.

wxb1ank commented:
Thanks for clarifying, @yume-chan. As I was exploring the VS Code API I noticed that deprecation, but I didn't realize what exactly the cause for the path error was.

At the end of the day, the simplest solution would be to rollback my commits and replace `path` with `fsPath`; that would at least accomplish the original goal of this PR. However, looking through the VS Code extension code, I think the way we're using raw strings to represent paths and importing the Node `fs` module for FS operations is a bit messy. At the same time, I don't know if replacing `fs` with `vscode.FileSystem` will be a regression in this case. Perhaps a wrapper that uses the VS Code API as a base and draws from Node when necessary is warranted?

yume-chan commented:
I thought `vscode.FileSystem` was added for Language Server type extensions to support virtual file systems, but the extension global store in this case should always be on the local disk, so Node.js `fs` module will work.

However when reading the documentation more carefully, I found this:

> \> **globalStorageUri: Uri**
>
> [...]
>
> * see - `workspace.fs` for how to read and write files and folders from an uri.
>
> -- https://code.visualstudio.com/api/references/vscode-api#ExtensionContext.globalStorageUri

So maybe Code expects extensions to use `vscode.FileSystem` to operate on `globalStorageUri`.

wxb1ank commented:
Personally, I think `Uri.fsPath` is still a 'correct' way to do this, documentation wise. I found this in `Uri.fsPath`:

> The resulting string shall not be used for display purposes but for disk operations, like readFile et al.

I didn't previously consider your comment about virtual file systems, but seeing as `globalState` is always local, I think this solution is fine for now.

matklad commented:
I've pushed a fix to *just* use fsPath, which fixed the immediate issue. However, switching from fs to vscode.fs is something we need to do as well, so it's best to rebase this PR on master!

Thanks a bunch for figuring this all out @wxb1ank and @yume-chan, it was instrumental in preparing the quick fix for todays release!

wxb1ank commented:
I found a vscode-python extension that documents their transition from Node's `fs` to `vscode.FileSystem`: https://github.com/microsoft/vscode-python/issues/8533. I think this proves that the latter can (mostly) replace the former. I will rebase and look into this later this week.

wxb1ank commented:
Sorry it took me a bit. Coming back to this, it looks like only `net.ts` uses the `fs` module now. Unfortunately, I still don't see a good solution to this: AFAIK, `vscode.FileSystem` can't modify file permissions (.e.g., make the downloaded executable `rx`), nor does it provide a suitable alternative to `fs.createWriteStream()` (necessary for storing gunzip streams*).

I've linted and rebased, and I think this is ready for review again.

*We also need a way to store non-gunzipped (gzipped?) streams. `node-fetch` conveniently provides `Response.buffer()` for this.

matklad commented:
bors r+

Let's see if this breaks anything :D

lnicola commented on editors/code/src/toolchain.ts:
```diff
@@ -158,9 +157,9 @@ export const getPathForExecutable = memoize(
+            const standardPath = vscode.Uri.joinPath(vscode.Uri.file(os.homedir()), ".cargo", "bin", executableName);
 
-            if (isFile(standardPath)) return standardPath;
+            if (isFile(standardPath.path)) return standardPath.path;
```
We convert an `Uri` to a string here, then convert it back to an `Uri`. And I'm not sure what happens if the path doesn't exist -- is there a way to `require("vscode")` in the Developer Tools?

bjorn3 replied:
> is there a way to require("vscode") in the Developer Tools?

I would expect not. The developer tools run inside the chromium based part of electron. The extensions run in the node.js based part.

lnicola replied:
@wxb1ank I haven't looked at the code outside of the diff, but do you think this can be simplified?

lnicola commented on editors/code/src/toolchain.ts:
```diff
@@ -181,12 +180,6 @@ function lookupInPath(exec: string): boolean {
-        return false;
-    }
+async function isFile(path: string): Promise<boolean> {
+    return ((await vscode.workspace.fs.stat(vscode.Uri.file(path))).type & vscode.FileType.File) !== 0;
```
What happens here if the file doesn't exist?

wxb1ank replied:
If I had to guess... [it throws](https://github.com/microsoft/vscode/blob/60d2aeee4b1c3104bd55c99679546168df4f6001/src/vs/vscode.d.ts#L7141). (Note: that's for `FileSystemProvider`, not `FileSystem`, but I'm assuming the latter is a wrapper over the former?) This should definitely be wrapped in a try/catch block ASAP.

I'm not very proud of my implementation for this function, either; I wanted to maintain the original function signature (i.e. with a string path, not `Uri`) so that it could be reused in `lookupInPath()`.

Looking at it now, I think it would be best to separate `isFile()` into two functions: one that takes a string path, and one that takes a `Uri`. The former should be a simple wrapper over the latter, translating the string into a `Uri` with `Uri.file(path)`.

This PR has already been merged, so tomorrow (later today?) I can try and clean this up in a new one.
